### PR TITLE
WIP: Add dask backend

### PR DESCRIPTION
**Pull request by gerrymanoim**
This is a very WIP dask backend for ibis. Related issue: https://github.com/ibis-project/ibis/issues/2245. 

It is very closely modeled in the existing pandas backend. At this point a large percentage of the test suite passes (I've tried to `pytest.mark.xfail` things I know are WIP), though there are some large outstanding issues (listed below). 

Before spending more time tackling the issues below, I wanted to check in the the ibis maintainers:

1. Is this still a backend of interest that would be accepted in a PR?
2. All the broken functionality below notwithstanding, does this approach seem reasonable at a high level?

To be clear - I don't believe this PR is ready for a full code review (there's a lot of fixing and cleanup I want to do/know I need to do), I'm just looking for some early feedback on approach and acceptability. 

# Things I know are broken

## Broken in execution

- Windowing
- Interpolation for quantile
- Better join performance
- fix implementation of execute_materialized_join
- execute_substring_series_series
- execute_node_struct_field_series_group_by
- Any aggregation involving `SeriesGroupBy` is fairly broken
  - execute_binary_op_series_group_by
  - execute_binary_op_series_gb_simple
- execute_aggregation_dataframe
- execute_selection_dataframe and compute_sorted_frame are fairly broken

## Broken elsewhere

- create_table in client.py

## Also clean up

- various leftover todo items
- There's a lot of warnings coming from dask about unspecified meta, we can clean that up. 
- Leftover bad calls directly to `dd.Series` and `dd.DataFrame`, littered throughout the code

### CODE REVIEW

**Comment by jreback** [comment]
thanks @gerrymanoim 

> Is this still a backend of interest that would be accepted in a PR?
absolutely!

> All the broken functionality below notwithstanding, does this approach seem reasonable at a high level?
haven't looked  yet, but if you have modeled after the pandas backend then yes this would be reasonable. with the exception of course that dask produces a graph that is itself deferred, then we call execute on it (unlike pandas where we actually execute directly); so this is more similar to the compile backends in that regards.

**Comment by datapythonista** [comment]
Thanks for the work on this @gerrymanoim, great contribution.

@jreback, I think it would be better to add new backends into different repos. Once #2379 is finished, there won't be any difference for users where the backend is. And with probably around 20 backends soon, I think being able to give commit rights to people to specific backends is a clear advantage. Does it sound good to you for the dask backend?

**Comment by jreback** [comment]
@datapythonista yeah not averse to doing this in a separate repo; the infra will likley be in place before this is merged anyhow.

**Comment by datapythonista** [comment]
@gerrymanoim do you mind reopening this PR to this other repo please: https://github.com/ibis-project/ibis-dask

**Comment by gerrymanoim** [comment]
Yep - will reopen there.

**Comment by gerrymanoim** [comment]
@datapythonista - Can you confirm that this is the expected package structure for backends (I want to make sure I'm reading https://github.com/ibis-project/ibis/pull/2379/files#diff-5806531ba2b4ac755e41f7776e08d3ed correctly). 

```
ibis-dask/
├── README.md
├── TODO.md
├── ibis-dask
│   ├── __init__.py
│   ├── aggcontext.py
│   ├── api.py
│   ├── client.py
│   ├── core.py
│   ├── dispatch.py
│   ├── dispatcher.py
│   ├── execution
│   ├── tests
│   ├── trace.py
│   └── udf.py
└── setup.py
```

**Comment by datapythonista** [comment]
Looks good to me. Just one comment compared to what we've got in most backends now. I wouldn't have the `api.py` file, and use `__init__.py` instead. I think that's the Pythonic way of defining where is the API.

You may also want to have an `environment.yml` file with the dependencies (including dependencies for development). And also set up GitHub actions to check linting, mypy, run the backend tests...

Thanks!

**Comment by datapythonista** [comment]
Reopening this, see https://github.com/ibis-project/ibis-dask/issues/3#issuecomment-701375582

@gerrymanoim do you mind moving the backend to `backends/dask`, and making the entry point `__init__.py` instead of `api.py` please? We're moving all backends to that structure now.

If you can also merge master, and we'll review the code then. Thanks for you work on this, and sorry for all the changes.

**Comment by icexelloss** [comment]
@gerrymanoim Thanks for working on this! I'd like to review this PR. Not sure what status it currently is but can you please notify when this is ready for review?

**Comment by gerrymanoim** [comment]
@icexelloss - Sure, will do. Right now it is fairly WIP and definitely needs some code clean up (even if I don't yet go fix all the broken pieces and raise `NotImplementedError`s for now, I just haven't had a good chunk of time to dive into this again.

**Comment by datapythonista** [comment]
Since there is not much activity here, I think it's better to go back to the original approach, and start this in a separate repo, and move to the Ibis repo later on, when it makes sense. @gerrymanoim feel free to ask any of us for reviews, even if the code is in a repo in your username initially.

**Comment by jreback** [comment]
no this is fine as is
splitting off to a separate repo is a non starter for a backend like this

**Comment by datapythonista** [comment]
What are the next steps with this then? It's been inactive for one month and a half.

**Comment by jreback** [comment]
@gerrymanoim is going to come back to this soon

**Comment by gerrymanoim** [comment]
Hi - so sorry for the delay on this. I had  a bunch of work stuff come up (https://www.quantopian.com/posts/quantopians-community-services-are-closing) that took up the majority of my time. I'm working on ramping back up on this PR this/next week now that things have quieted down.

File: ibis/backends/dask/__init__.py
Comment (jreback) [inline ibis/backends/dask/__init__.py:65]: ideally put things like this else where *maybe* dask/expr.py or similiar

File: ibis/backends/dask/client.py
Comment (jreback) [inline ibis/backends/dask/client.py:61]: hmm i think we have this elsewhere no?

File: ibis/backends/dask/client.py
Comment (jreback) [inline ibis/backends/dask/client.py:125]: can we simplify (e.g. no python 2) or is that just copied?

File: ibis/backends/dask/client.py
Comment (jreback) [inline ibis/backends/dask/client.py:246]: hmm, where did this come from? we really want to let Timestamp handle this and not do these types of conversions directly.

File: ibis/backends/dask/client.py
Comment (jreback) [inline ibis/backends/dask/client.py:255]: most things in this module should be in dtypes.py or types.py

File: ibis/backends/dask/client.py
Comment (jreback) [inline ibis/backends/dask/client.py:346]: except these of course!

File: ibis/backends/dask/core.py
Comment (jreback) [inline ibis/backends/dask/core.py:20]: hmm this is pretty misleading, dask *does* compile to a dask expression, and then *execute* computes on it, unless pandas which doesn't have a compliation step at all. this is much more akin to the way spark works, no?

File: ibis/backends/dask/dispatcher.py
Comment (jreback) [inline ibis/backends/dask/dispatcher.py:3]: TODO item, can consolidate this with the code in pandas and put in a common location (e.g. that's a good pre-cursor PR to bite off here).

File: ibis/backends/dask/execution/__init__.py
Comment (jreback) [inline ibis/backends/dask/execution/__init__.py:8]: yeah don't really love this pattern but ok

**Review by jreback — CHANGES_REQUESTED** [review]
just had a very quick look. nice docs! i think it might pay to pick some of this off and do multiple pre-cursors PRs, e.g.

- push dask with almost a trivial import (like no code and a single test) to get the CI working
- factor out dispatcher stuff
- initial implmentation with everything working (except for skips)
- adding working things one by one as they work (e.g. the things you mention in the TODO), e.g. windows can wait. 

This makes reviewing *way* easier and way faster.

**Comment by gerrymanoim** [comment]
Thanks for having a look @jreback - I'll address your comments individually above. 

>  i think it might pay to pick some of this off and do multiple pre-cursors PR

Do you prefer the PRs to be against this PR or against master? Happy to change this to whatever makes this convenient to review. It is a ton of code, but it largely ports the pandas backend directly where possible (I still need to fix up docs to correct all of them). There's also areas for code reuse that I'd like to refactor. 

As a general status, most things work (or are TODO skipped) except for `aggcontext` and `window` .

  - Aggregations are different enough in Dask that anything accessing the `GroupBy` object directly needs to be changed. I've written custom aggregations in a few places so far, but there's more work to be done. 
  - Dask has no good `sort_values` mechanism (AFAICT) which makes windowing hard. It seems that your options are to make the column into an index (which is expensive and means you can't sort on multiple cols), but I wanted to dig into this more before doing a lot of work here. 

I've currently picked up work again on fixing some functions I previously xfailed (`execute_materialized_join`, `execute_substring_series`, etc.). 

Full results from a recent local run:

```
ibis/backends/dask/tests/test_aggcontext.py FFF                                                                                                      [  0%]
ibis/backends/dask/tests/test_client.py ...x....xxxxxxxx.xxx                                                                                         [  3%]
ibis/backends/dask/tests/test_core.py ...........                                                                                                    [  4%]
ibis/backends/dask/tests/test_datatypes.py ...................................x...............                                                       [ 11%]
ibis/backends/dask/tests/test_dispatcher.py ......................                                                                                   [ 14%]
ibis/backends/dask/tests/test_schema.py ..................                                                                                           [ 16%]
ibis/backends/dask/tests/test_udf.py ....x...x......xxxx..xx.                                                                                        [ 19%]
ibis/backends/dask/tests/execution/test_arrays.py ...xx.......xxx.......xxx.................................                                         [ 27%]
ibis/backends/dask/tests/execution/test_cast.py ..........................x....x....x....x...x...x...........                                        [ 35%]
ibis/backends/dask/tests/execution/test_functions.py ..............................................................                                  [ 44%]
ibis/backends/dask/tests/execution/test_join.py ....xx.....xx.....xx....xx....xxxxxxxx....xx....xx....xx....xxxxxxxxxxxxxx....xxs....xxxxxxxxxxxxxxx [ 57%]
                                                                                                                                                     [ 57%]
ibis/backends/dask/tests/execution/test_maps.py .......                                                                                              [ 58%]
ibis/backends/dask/tests/execution/test_operations.py .....xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx................................xxxxxxxxxxxx. [ 71%]
...x..x............x..............xxxxxxxxxxxxxx...s.sssss.s.s.....x.                                                                                [ 80%]
ibis/backends/dask/tests/execution/test_strings.py .........................................                                                         [ 85%]
ibis/backends/dask/tests/execution/test_structs.py ...x                                                                                              [ 86%]
ibis/backends/dask/tests/execution/test_temporal.py .....................................                                                            [ 91%]
ibis/backends/dask/tests/execution/test_timecontext.py ....xxxxxx                                                                                    [ 92%]
ibis/backends/dask/tests/execution/test_window.py FFFFFFFFFFFFFFFFFFFFFFFFFFFsssssssssssssssssFFFFFFFFFFF                                            [100%
````

File: ibis/backends/dask/__init__.py
Comment (gerrymanoim) [reply ibis/backends/dask/__init__.py:65]: Should the pandas backend functions be moved to a `pandas/expr.py` as well? Or is there a good place to start putting functions that are directly shared between dask and pandas? I could import this directly from the pandas backend actually.

File: ibis/backends/dask/client.py
Comment (gerrymanoim) [reply ibis/backends/dask/client.py:61]: Another thing that is directly cloned from the pandas backend. I can change to import from there?

File: ibis/backends/dask/client.py
Comment (gerrymanoim) [reply ibis/backends/dask/client.py:125]: This is just copied (from https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/client.py#L120) - I can update it in both places.

File: ibis/backends/dask/client.py
Comment (gerrymanoim) [reply ibis/backends/dask/client.py:246]: Similarly: https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/client.py#L241

File: ibis/backends/dask/core.py
Comment (gerrymanoim) [reply ibis/backends/dask/core.py:20]: I can take a look at what pyspark does. 

I think the point I was trying to make here is that ibis-dask will ultimately pass you back a dask expression that you can either `.compute()` (giving you a `pd.DataFrame`) or continue to run computations on. So we're using ibis to build up the expression and it is ultimately up the user to compute that expression, rather than doing it for them.

File: ibis/backends/dask/__init__.py
Comment (jreback) [reply ibis/backends/dask/__init__.py:65]: so yes and yes. however these should be done independently. (e.g. separate PRs for pandas). may make sense to move to a common location (or dask can simply import the pandas ones)

File: ibis/backends/dask/client.py
Comment (jreback) [reply ibis/backends/dask/client.py:61]: yes, go ahead an import things from pandas directly (if its more of 'shared' code then we can likley find a better place), but easy enough to just import now.

File: ibis/backends/dask/client.py
Comment (jreback) [reply ibis/backends/dask/client.py:125]: :->

File: ibis/backends/dask/client.py
Comment (jreback) [reply ibis/backends/dask/client.py:246]: ok ic, yeah you can likely just import things like this

File: ibis/backends/dask/core.py
Comment (jreback) [reply ibis/backends/dask/core.py:20]: yes that's exactly right, i think we just want to explain that in the documentation (e.g. here) 'execute ibis expressions directly' is true, but a bit misleading (e.g. say what you wrote right above)

**Comment by jreback** [comment]
>  Do you prefer the PRs to be against this PR or against master? Happy to change this to whatever makes this convenient to review. It is a ton of code, but it largely ports the pandas backend directly where possible (I still need to fix up docs to correct all of them).

if you can split things off into separable chunks would prefer that. 10 - 1000 PRs are much better than 1 10k

> There's also areas for code reuse that I'd like to refactor. Do you have a preference on creating a shared location for pandas and dask to both pull code from or for the dask backend to pull from the pandas backend when needed?

so happy either way here, e.g. if we find that having a common location is helpful then let's do that. it may that its easier to *copy* code somewhat, then later consolidate. again PRs to this affect that are focused and do one thing are the best.

Ideal is to have straightforward, easy to review PRs !) , then we can have a good velocity and build on them. I find that larger PRs just get bogged down and frustrate all sides.

**Comment by gerrymanoim** [comment]
Okay - I'll reopen new smaller PRs chunked as you suggested: 

1. Add a trivial skeleton + CI
2. Pull the dispatcher things from the pandas backend and the rest of the supporting files outside /execution/ (client, core, trace), importing from pandas where possible. 
3. Known working-ish pieces (pieces in `/execution/`). 
4. TODOs for windowing, aggregations, etc

**Pull request closed without merging**
